src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp

```diff
@@ -5338,22 +5338,41 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
   } else {
     Label NEXT_DWORD, SHORT, TAIL, TAIL2, STUB,
         CSET_EQ, LAST_CHECK;
+    // When the base is not aligned to 8 bytes, then we let
+    // the compare loop include the array length, and skip
+    // the explicit comparison of length.
+    bool is_8aligned = is_aligned(base_offset, BytesPerWord);
     mov(result, false);
     cbz(a1, DONE);
     ldrw(cnt1, Address(a1, length_offset));
     cbz(a2, DONE);
     ldrw(cnt2, Address(a2, length_offset));
+    if (!is_8aligned) {
+      // Increase loop counter by size of length field.
+      addw(cnt1, cnt1, BytesPerInt / elem_size);
+      addw(cnt2, cnt2, BytesPerInt / elem_size);
+    }
     // on most CPUs a2 is still "locked"(surprisingly) in ldrw and it's
     // faster to perform another branch before comparing a1 and a2
     cmp(cnt1, (u1)elem_per_word);
     br(LE, SHORT); // short or same
-    ldr(tmp3, Address(pre(a1, base_offset)));
+    if (is_8aligned) {
+      ldr(tmp3, Address(pre(a1, base_offset)));
+    } else {
+      ldr(tmp3, Address(pre(a1, length_offset)));
+    }
```

Comment:
Something like this could avoid the ifs. Also the `const int start_offset = is_8aligned ? base_offset : length_offset;` could be pushed up and shared for the simple implementation, which probably should be fixed as well.
```suggestion
    const int start_offset = is_8aligned ? base_offset : length_offset;
    ldr(tmp3, Address(pre(a1, start_offset)));
```